### ISSUE TITLE
Authenticating reverse proxy question

### ISSUE THREAD

**Issue by CJHarmath** [Bikeman868/UrlRewrite.Net#17]
Just saw this module linked at the [Microsoft IIS rewrite module 2.1 announcement page](https://blogs.iis.net/iisteam/url-rewrite-v2-1) and were wondering if anyone had setup an authenticating reverse proxy which sets the authenticated user name ?
i.e.: to front services like Jenkins or Splunk but with for example Kerberos authentication happening at IIS then set the user name as a header and forward to the backend.

**use case** 
you are in an AD environment with working Kerberos setup and would want to rely on that to provide single sign-on for apps like Jenkins or Splunk via an authenticated IIS based reverse proxy which does auth, sets header then reverse proxies.

Jenkins for example have a plugin for allowing it to be reverse proxied 
https://plugins.jenkins.io/reverse-proxy-auth-plugin
Which requires the X-Forwarded-User header to be set to the user, but I wonder how is anyone doing that as the rewrite modules are executing before authentication happens ?

Just started looking at the source, so apologies if this is already covered and I am just missing something.
So far have only looked at the Microsoft rewrite module which seems to be executing before auth can happen, so that's no value for me it seems.
I think this topic is worth discussing as I am probably not the only one with this use case in mind.

Thank you!

**Comment by CJHarmath** [comment]
I thing I had in mind was to switch event handler in the Init method
from:
`application.BeginRequest += OnBeginRequest;`
to:
`application.PostAuthenticateRequest += OnPostAuthenticateRequest;`

As only after that event you can grab the username, but then I wanted to ask here first to discuss the options.

**Comment by Bikeman868** [comment]
Thanks for your comments and suggestions. When I originally started this project it was because the application I inherited had more than 2000 rewrite rules and this delayed processing of every request by 10s of milliseconds.

At that time my objective was to make the rule processing much more efficient, and also provide a nested hierarchy of rules so that many of my 2000 rules could be skipped over for a given request.

At that time I was not aware that the Microsoft implementation had the ability to perform reverse proxy, and even now I have never used this functionality, and an not 100% sure what it does.

I have had a couple of other requests for adding a reverse proxy capability, but a) I don't need it myself b) I don't know how to write the requirements because I don't fully understand what it is trying to do, and c) I have several open source projects on the go, so conflicting time pressures.

If you know what a reverse proxy feature should do, or how it should work, then maybe you can write the requirements. If you have the time/inclination to implement this feature I am willing to accept your pull request (after careful review of course).

**Comment by Bikeman868** [comment]
So, I did a bit of reading about reverse proxy and it is basically a load balancer. I am not sure why the Microsoft URL rewriter contains any functionality in this area which seems relatively unrelated to URL rewriting.

I think that if you want a load balancer then you should get a load balancer, I am not clear why anyone would want to build that functionality into their application.

**Comment by Bikeman868** [comment]
For anyone following this thread you might be interested to know that I have built an open-source load-balancer called "Gravity". One of the available processing nodes that can be configured implements these rewriting rules - I copied the code directly over so it's fully compatible.
Check out my "Gravity" repo.
